pallets/gear/src/builtin.rs

```diff
@@ -0,0 +1,70 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use gear_core::ids::{BuiltinId, ProgramId};
+
+/// A trait representing a registry that provides methods to lookup a builtin actor
+pub trait BuiltinRouter {
+    type QueuedDispatch;
+    type Output;
+
+    /// Looks up a builtin actor by its actor id.
+    fn lookup(&self, id: &ProgramId) -> Option<BuiltinId>;
```

Comment:
why do you need BuiltinId if you already have ProgramId?
Maybe just true/false?

Reply:
A message address is a `ProgramId`, but in order to find the correct actor who is going to process it we must "reverse-map" it to a `BuiltinId` since builtin actors only know their own `BuiltinId`.

Reply:
Another update: now we actually do need a value as the output of the `lookup` in order to enable the check whether a dispatch can be handled by a builtin actor before consuming it to avoid 
[excessive cloning of a dispatch](https://github.com/gear-tech/gear/pull/3624/files/6f271facf645b83cbab4ac57b5c24895cdfee62a#diff-25fc00efe7529b336fe02426d3e7fab7ab580dbb076b03e9e8b17cf632d2c35e).